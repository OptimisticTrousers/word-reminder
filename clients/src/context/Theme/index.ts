export { ThemeContext } from "./Context";
export { ThemeProvider } from "./Provider";
