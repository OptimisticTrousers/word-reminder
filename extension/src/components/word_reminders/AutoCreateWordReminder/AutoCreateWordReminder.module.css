.auto-word-reminder,
button.create {
  background-color: var(--card-bkg-color);
  padding: 1rem;
  border-radius: 8px;
  border: 1px solid var(--border-color);
}

button.create:hover {
  background-color: var(--hover-bkg-color);
}

.create__container {
  display: flex;
  align-items: center;
}

.create__icon {
  width: 48px;
  height: 48px;
  margin-right: 12px;
  flex-shrink: 0;
}

.create__text {
  text-align: start;
}

.create__title {
  font-size: 1.125rem;
  line-height: 1.75rem;
  font-weight: 600;
}

.create__description {
  line-height: 1.5rem;
  color: var(--secondary-text-color);
}

.auto-word-reminder__buttons {
  margin-top: 0.75rem;
  display: flex;
  gap: 0.75rem;
}

.auto-word-reminder__button {
  flex: 1;
}
