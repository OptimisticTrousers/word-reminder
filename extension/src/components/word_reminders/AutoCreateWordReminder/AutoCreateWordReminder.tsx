import { CirclePlus } from "lucide-react";
import { useState } from "react";
import cronstrue from "cronstrue";
import CSSModules from "react-css-modules";

import styles from "./AutoCreateWordReminder.module.css";
import { User } from "common";
import { useOutletContext } from "react-router-dom";
import { CreateAutoWordReminderModal } from "../../modals/CreateAutoWordReminderModal";
import { autoWordReminderService } from "../../../services/auto_word_reminder_service/auto_word_reminder_service";
import { useQuery } from "@tanstack/react-query";
import { Loading } from "../../ui/Loading";
import { UpdateAutoWordReminderModal } from "../../modals/UpdateAutoWordReminderModal";
import { DeleteAutoWordReminderModal } from "../../modals/DeleteAutoWordReminderModal";
import { msToUnits } from "../../../utils/date/date";
import { ErrorMessage } from "../../ui/ErrorMessage";

export const AutoCreateWordReminder = CSSModules(
  function () {
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
    const [isCreateModalOpen, setIsCreateModalOpen] = useState(false);
    const [isUpdateModalOpen, setIsUpdateModalOpen] = useState(false);

    const { data, failureReason, isLoading } = useQuery({
      queryKey: ["autoWordReminders"],
      queryFn: () => {
        return autoWordReminderService.getAutoWordReminder({ userId });
      },
    });

    const toggleCreateModal = () => {
      setIsCreateModalOpen((prevValue) => !prevValue);
    };
    const toggleUpdateModal = () => {
      setIsUpdateModalOpen((prevValue) => !prevValue);
    };

    const toggleDeleteModal = () => {
      setIsDeleteModalOpen((prevValue) => !prevValue);
    };

    if (failureReason) {
      return <ErrorMessage message={failureReason.message} />;
    }

    if (isLoading) {
      return <Loading />;
    }

    const autoWordReminder = data?.json.autoWordReminder;

    if (!autoWordReminder) {
      return (
        <>
          {isCreateModalOpen && (
            <CreateAutoWordReminderModal toggleModal={toggleCreateModal} />
          )}
          <button
            styleName="create"
            onClick={toggleCreateModal}
            aria-haspopup="dialog"
            aria-labelledby="title"
          >
            <div styleName="create__container">
              <CirclePlus styleName="create__icon" />
              <div styleName="create__text">
                <h2 styleName="create__title" id="title">
                  Create Auto Word Reminder
                </h2>
                <p styleName="create__description">
                  Automatically create word reminder on an interval to remember
                  words you come across in your readings. Set it and forget it.
                </p>
              </div>
            </div>
          </button>
        </>
      );
    }

    const reminder = cronstrue.toString(autoWordReminder.reminder);

    const duration = msToUnits(autoWordReminder.duration);

    return (
      <>
        <section styleName="auto-word-reminder">
          <h3 styleName="auto-word-reminder__heading">Auto Word Reminder</h3>
          <div styleName="auto-word-reminder__id">
            ID: {autoWordReminder.id}
          </div>
          <div styleName="auto-word-reminder__reminder">
            Reminder: {reminder}
          </div>
          <div styleName="auto-word-reminder__duration">
            Weeks: {duration.weeks}, Days: {duration.days}, Hours:{" "}
            {duration.hours}, Minutes: {duration.minutes}
          </div>
          <div styleName="auto-word-reminder__is-active">
            Active (whether the word reminder will actively remind you of the
            words in it): {autoWordReminder.is_active ? "Yes" : "No"}
          </div>
          <div styleName="auto-word-reminder__has_reminder_onload">
            Has Reminder Onload (reminds you of these words when you open your
            browser): {autoWordReminder.has_reminder_onload ? "Yes" : "No"}
          </div>
          <div styleName="auto-word-reminder__has_learned_words">
            Has Learned (whether to include words that you have already
            learned): {autoWordReminder.has_learned_words ? "Yes" : "No"}
          </div>
          <div styleName="auto-word-reminder__has-reminder-onload">
            Sort Mode (determines sort mode): {autoWordReminder.sort_mode}
          </div>
          <div styleName="auto-word-reminder__has-reminder-onload">
            Word Count (how many words to include):{" "}
            {autoWordReminder.word_count}
          </div>
          <div styleName="auto-word-reminder__created_at">
            This auto word reminder was created on{" "}
            {autoWordReminder.created_at.toLocaleString()}
          </div>
          <div styleName="auto-word-reminder__updated_at">
            This auto word reminder was updated on{" "}
            {autoWordReminder.updated_at.toLocaleString()}
          </div>
          <div styleName="auto-word-reminder__buttons">
            <button
              styleName="auto-word-reminder__button"
              onClick={toggleUpdateModal}
            >
              Update
            </button>
            <button
              styleName="auto-word-reminder__button"
              onClick={toggleDeleteModal}
            >
              Delete
            </button>
          </div>
        </section>
        {isUpdateModalOpen && (
          <UpdateAutoWordReminderModal
            autoWordReminder={autoWordReminder}
            toggleModal={toggleUpdateModal}
          />
        )}
        {isDeleteModalOpen && (
          <DeleteAutoWordReminderModal
            autoWordReminderId={autoWordReminder.id}
            toggleModal={toggleDeleteModal}
          />
        )}
      </>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
