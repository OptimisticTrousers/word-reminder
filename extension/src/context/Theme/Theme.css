.container {
  background-color: var(--bkg-color);
  color: var(--text-color);
}

.container--light {
  --bkg-color: #fff;
  --hover-bkg-color: rgb(12, 11, 9);
  --text-color: #222;
  --anchor-color: #0033cc;
  --input-bkg-color: #fff;
  --label-color: none;
  --shadow-color: none;
  --input-color: none;
  --primary-card-bkg-color: none;
  --secondary-card-bkg-color: none;
  --accordion-bkg-color: none;
  --control-bkg-color: none;
  --border-color: none;
  --icon-color: none;
  --navigation-bkg-color: none;
  --bottom-navigation-bkg-color: none;
  --active-navigation-bkg-color: none;
  --modal-bkg-color: rgba(245, 245, 245, 0.8);
  --break-color: #dedede;
  --secondary-icon-color: #606770;
  --hover-icon-color: rgb(59, 130, 246);
  --alert-text-color: rgb(185, 28, 28);
  --alert-bkg-color: rgb(254, 226, 226);
  --white-text-color: rgb(31, 41, 55);
  --white-bkg-color: rgb(243, 244, 246);
  --hover-white-bkg-color: rgb(229, 231, 235);
  --red-bkg-color: rgb(239, 68, 68);
  --hover-red-bkg-color: rgb(220, 38, 38);
}

.container--dark {
  --bkg-color: #242424;
  --text-color: #eee;
  --secondary-text-color: rgb(107, 114, 128);
  --tertiary-text-color: rgb(175, 170, 156);
  --anchor-color: #809fff;
  --secondary-anchor-color: #181819;
  --input-bkg-color: #2a2a2a;
  --label-color: #bac0ce;
  --shadow-color: #757575;
  --input-color: #a3a3a3;
  --card-bkg-color: #2a2a2a;
  --secondary-card-bkg-color: #161d3a;
  --accordion-bkg-color: #1f1e22;
  --top-navigation-bkg-color: #010001;
  --control-bkg-color: #2f343d;
  --border-color: #333;
  --icon-color: #efefef;
  --bottom-navigation-bkg-color: #f8f8f8;
  --active-navigation-bkg-color: #d3d3d3;
  --modal-bkg-color: rgba(42, 42, 42, 0.8);
  --break-color: #dedede;
  --secondary-icon-color: #606770;
  --hover-icon-color: rgb(59, 130, 246);
  --alert-text-color: rgb(185, 28, 28);
  --alert-bkg-color: rgb(254, 226, 226);
  --white-text-color: rgb(31, 41, 55);
  --white-bkg-color: rgb(243, 244, 246);
  --hover-white-bkg-color: rgb(229, 231, 235);
  --red-bkg-color: rgb(239, 68, 68);
  --hover-red-bkg-color: rgb(220, 38, 38);

  /* --color-control-bg: #2f343d;
  --color-word-bg: #161d3a;
  --color-base-text: white;
  --color-input-bg: #fff;
  --color-input-text: #a3a3a3;
  --color-button-bg: #7f8ff4;
  --color-button-text: #fff;
  --color-options-text: black;
  --color-accordion-bg: #1f1e22;
  --color-word-text: white;
  --color-accordion-text: white;
  --color-shadow: #757575;
  --color-label-text: #bac0ce; */
}

.container a {
  color: var(--anchor-color);
}

.container button,
.container input[type="button"] {
  padding: 0.6rem;
  border-radius: 8px;
  border: none;
  background-color: var(--primary-bkg-color);
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.container button:hover,
.container input:hover[type="button"] {
  background-color: var(--primary-bkg-hover-color);
}

.container button:focus,
.container input:focus,
.container select:focus {
  outline: none;
  box-shadow: 0 0 5px var(--primary-color);
}

.container img {
  filter: brightness(0.8) contrast(1.2);
}

.container input {
  border: 1px solid var(--border-color);
}

.container input,
.container select,
.container fieldset {
  padding: 0.5rem;
  border-radius: 8px;
  background-color: var(--input-bkg-color);
}
