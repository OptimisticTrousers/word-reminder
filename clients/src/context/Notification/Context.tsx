import { createContext } from "react";

import { NotificationContext as Context } from "./Provider";

export enum NOTIFICATION_ACTIONS {
  SUCCESS = "success",
  ERROR = "error",
}

export const NotificationContext = createContext<Context>({
  showNotification: function () {},
  dismissNotification: function () {},
});
