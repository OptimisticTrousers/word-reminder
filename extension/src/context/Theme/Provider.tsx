import { Dispatch, ReactNode, SetStateAction } from "react";

import { Theme, ThemeContext } from "./Context";
import { useLocalStorage } from "../../hooks/useLocalStorage";

interface Props {
  children: ReactNode;
}

export interface ThemeContext {
  theme: Theme;
  setTheme: Dispatch<SetStateAction<Theme>>;
}

export function ThemeProvider({ children }: Props) {
  const [theme, setTheme] = useLocalStorage("theme", Theme.Dark);

  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
}
