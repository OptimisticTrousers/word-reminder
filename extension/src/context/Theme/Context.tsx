import { createContext } from "react";

import { ThemeContext as Context } from "./Provider";

export enum Theme {
  Dark = "dark",
  Light = "light",
}

export const ThemeContext = createContext<Context>({
  theme: Theme.Dark,
  setTheme: function () {},
});
