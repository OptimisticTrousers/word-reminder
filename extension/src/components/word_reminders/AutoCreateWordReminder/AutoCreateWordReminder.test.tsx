import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AutoCreateWordReminder } from "./AutoCreateWordReminder";
import { Theme, ThemeContext } from "../../../context/Theme/Context";
import { createRoutesStub, Outlet } from "react-router-dom";

vi.mock(
  "../../modals/AutoCreateWordReminderModal/AutoCreateWordReminderModal",
  () => {
    return {
      AutoCreateWordReminderModal: function ({
        toggleModal,
      }: {
        toggleModal: () => void;
      }) {
        return (
          <div role="dialog">
            <button onClick={toggleModal}>Close</button>
          </div>
        );
      },
    };
  }
);

describe("AutoCreateWordReminder component", () => {
  const searchParams = new URLSearchParams();
  function setup(testUser: { id: string; auto: boolean }, theme: Theme) {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <ThemeContext.Provider value={{ theme, toggleTheme: vi.fn() }}>
                  <AutoCreateWordReminder searchParams={searchParams} />;
                </ThemeContext.Provider>
              );
            },
          },
        ],
      },
    ]);
    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/"]} />),
    };
  }

  describe("auto", () => {
    it("renders untoggled button", async () => {
      const testUser = { id: "1", auto: false };
      const { asFragment } = setup(testUser, Theme.Dark);

      const heading = screen.getByRole("heading", {
        name: "Automatic Create Word Reminder",
      });
      const description = screen.getByText(
        "Automatically create word reminder on an interval to remember words you come across in your readings. Set it and forget it."
      );
      expect(heading).toBeInTheDocument();
      expect(description).toBeInTheDocument();
      expect(asFragment).toMatchSnapshot();
    });

    it("renders toggled button", async () => {
      const testUser = { id: "1", auto: true };
      const { asFragment } = setup(testUser, Theme.Dark);

      const heading = screen.getByRole("heading", {
        name: "Automatic Create Word Reminder Toggled...",
      });
      const description = screen.getByText(
        "Automatically create word reminder on an interval to remember words you come across in your readings. Set it and forget it."
      );
      expect(heading).toBeInTheDocument();
      expect(description).toBeInTheDocument();
      expect(asFragment).toMatchSnapshot();
    });
  });

  it("has light class", () => {
    const testUser = { id: "1", auto: false };
    setup(testUser, Theme.Light);

    const button = screen.getByRole("button", {
      name: "Automatic Create Word Reminder",
    });
    expect(button.getAttribute("class")).toContain(`create--${Theme.Light}`);
  });

  it("has dark class", () => {
    const testUser = { id: "1", auto: false };
    setup(testUser, Theme.Dark);

    const button = screen.getByRole("button", {
      name: "Automatic Create Word Reminder",
    });
    expect(button.getAttribute("class")).toContain(`create--${Theme.Dark}`);
  });

  it("opens model when the create word reminder is clicked", async () => {
    const testUser = { id: "1", auto: false };
    const { user } = setup(testUser, Theme.Dark);

    const button = screen.getByRole("button", {
      name: "Automatic Create Word Reminder",
    });
    await user.click(button);

    const modal = screen.getByRole("dialog");
    expect(modal).toBeInTheDocument();
  });

  it("closes model when the create word reminder is clicked and modal close button is clicked", async () => {
    const testUser = { id: "1", auto: false };
    const { user } = setup(testUser, Theme.Dark);

    const button = screen.getByRole("button", {
      name: "Automatic Create Word Reminder",
    });
    await user.click(button);
    const closeModalButton = screen.getByRole("button", { name: "Close" });
    await user.click(closeModalButton);

    const modal = screen.queryByRole("dialog");
    expect(modal).not.toBeInTheDocument();
  });
});
