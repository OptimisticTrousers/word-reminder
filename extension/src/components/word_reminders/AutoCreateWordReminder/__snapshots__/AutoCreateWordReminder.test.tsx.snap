// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AutoCreateWordReminder component > auto > renders toggled button 1`] = `[Function]`;

exports[`AutoCreateWordReminder component > auto > renders untoggled button 1`] = `[Function]`;
